### PR TITLE
feat: new rule to detect enum value that do not respect specified type

### LINKED ISSUE #570: Create OpenAPI rule for enums that don't match the attribute type
As an example, if I have a numeric attribute in a JSON schema object. I want to ensure that any enumerations of that attribute are of the same numeric type (ie, not a string). It should be flagged as an error when this occurs.

For example, the JSON schema below should trigger an error:

```yaml
  a_model:
    type: object
    properties:
      id:
        type: integer
        description: Unique asset identifier
        enum:
          - 'a string!'
        format: int64
```

Since the `id` field is of type integer but the enum is a string. While this:

```yaml
  a_model:
    type: object
    properties:
      id:
        type: integer
        description: Unique asset identifier
        enum:
          - 123
        format: int64
```

Should not.

Created from https://github.com/stoplightio/studio/issues/116

### PR DESCRIPTION
**Pull request by nulltoken**
Fixes #570

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by nulltoken** [comment]
@chris-miaskowski Could you please take a look at this, please? It's not done yet, but this should be usable for a first round of tests.

If you could also run some local tests by checking this branch out and running `yarn cli lint...` against a couple files that would be amazing!

@P0lip I'd also be a taker for a review pass. FWIW, I foresee to enhance the testing of the function by throwing at it more combinations of formats and values.

File: src/functions/formatedEnum.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import { schema } from './schema';
+
+export const formatedEnum: IFunction = (targetVal, opts, paths, otherValues): void | IFunctionResult[] => {
+  const { enum: enumValues, ...innerSchema } = targetVal;
+  const schemaObject = { schema: innerSchema };
+
+  const incorrectValues: unknown[] = [];
+
+  // Maximum number of reported invalid enum values
+  const reportingThreshold = 3;
```
Comment (nulltoken) [inline src/functions/formatedEnum.ts]: The idea behind this is to not overburden the error message if the enum is pretty long and full of invalid values.

Optionally, this might be passed as a parameter to the function. Let me know if you think this add any value.

File: src/rulesets/oas/__tests__/formated-enums.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
+import { formatedEnum } from '../../../functions/formatedEnum';
+import { RuleType, Spectral } from '../../../index';
+import { rules } from '../index.json';
+
+describe('formated-enum', () => {
+  const s = new Spectral();
+  s.registerFormat('oas2', () => true);
```
Comment (nulltoken) [inline src/rulesets/oas/__tests__/formated-enums.ts]: Note to self: Add simple oas3 related tests

File: src/functions/formatedEnum.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import { IFunction, IFunctionResult } from '../types';
+import { schema } from './schema';
+
+export const formatedEnum: IFunction = (targetVal, opts, paths, otherValues): void | IFunctionResult[] => {
```
Comment (P0lip) [inline src/functions/formatedEnum.ts]: Typo :D
```suggestion
export const formattedEnum: IFunction = (targetVal, opts, paths, otherValues)=> {
```

File: src/functions/index.ts [hunk 4]
```diff
@@ -11,4 +11,5 @@ export const functions = {
   undefined: require('./undefined').undefined,
   xor: require('./xor').xor,
   unreferencedReusableObject: require('./unreferencedReusableObject').unreferencedReusableObject,
+  formatedEnum: require('./formatedEnum').formatedEnum,
```
Comment (P0lip) [inline src/functions/index.ts]:
```suggestion
  formattedEnum: require('./formattedEnum').formattedEnum,
```

File: src/functions/formatedEnum.ts (hunk 1 above)
Comment (P0lip) [reply src/functions/formatedEnum.ts]: Perhaps make it configurable?

File: src/functions/formatedEnum.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  const { enum: enumValues, ...innerSchema } = targetVal;
```
Comment (P0lip) [inline src/functions/formatedEnum.ts]: Would it make sense to validate the input?
It's quite an assumption that `enumValues` is an array.

Comment (nulltoken) [reply src/functions/formatedEnum.ts]: @P0lip You're right. lodah.isArray would work on this.

However, how should we handle this?
 - throw? This would be caught by the runner and issue a console.warn in stdout
 - bail? And hope this will be caught by an oas related schema rule.

File: src/functions/formatedEnum.ts (hunk 1 above)
Comment (nulltoken) [reply src/functions/formatedEnum.ts]: Ok. I'll make this configurable

File: src/functions/formatedEnum.ts (hunk 3 above)
Comment (nulltoken) [reply src/functions/formatedEnum.ts]: Fixed

File: src/functions/index.ts (hunk 4 above)
Comment (nulltoken) [reply src/functions/index.ts]: Fixed

File: src/functions/formatedEnum.ts (hunk 1 above)
Comment (nulltoken) [reply src/functions/formatedEnum.ts]: @P0lip Done. Does that fit your expectation?

File: src/functions/formatedEnum.ts (hunk 5 above)
Comment (nulltoken) [reply src/functions/formatedEnum.ts]: I went ahead and chose to ignore and bail out. @P0lip Thoughts?

**Comment by nulltoken** [comment]
Just pushed an update to ease the coverage of detection of invalid format usage

File: src/functions/__tests__/formattedEnum.test.ts [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+      expect(runFormattedEnum(schema, defaultReportingThreshold)).toEqual([
+        {
+          message: 'Some enum values do not respect the specified format "int64": "a string!", "and another one!".',
+        },
+      ]);
+    });
+  });
+
+  describe('formats', () => {
+    const testCases: Array<[string, string, unknown[], unknown]> = [
```
Comment (nulltoken) [inline src/functions/__tests__/formattedEnum.test.ts]: Note to self: Add more test coverage to cover the following fomats: `int32`, `int64`, `float`, `double`, `byte`, `date`, `time`, `date-time`, `uri`, `uri-reference`, `uri-template`, `email`, `hostname`, `ipv4`, `ipv6`, `regex`, `uuid`, `json-pointer`, `relative-json-pointer`

File: src/functions/formatedEnum.ts (hunk 1 above)
Comment (P0lip) [reply src/functions/formatedEnum.ts]: Yeah, looking good.

File: src/functions/formatedEnum.ts (hunk 5 above)
Comment (P0lip) [reply src/functions/formatedEnum.ts]: Could perhaps report enum is malformed?

File: src/functions/formattedEnum.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  targetVal,
+  opts,
+  paths,
+  otherValues,
+): void | IFunctionResult[] => {
+  assertValidOptions(opts);
+
+  const { enum: enumValues, ...innerSchema } = targetVal;
+
+  if (!isArray(enumValues)) {
```
Comment (P0lip) [inline src/functions/formattedEnum.ts]:
```suggestion
  if (!Array.isArray(enumValues)) {
```
let's use native :D

File: src/functions/formattedEnum.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
+import { AssertionError } from 'assert';
+import { isArray } from 'lodash';
```
Comment (P0lip) [inline src/functions/formattedEnum.ts]:
```suggestion
```

**Comment by philsturgeon** [comment]
This might actually go above and beyond. The feature request was for type, but this is type and format. Format is notoriously vague and I worry we're opening ourselves up to a maintainability nightmare, because JSON Schema 2019-09 has several new formats, and we're moving away from format being a validation keyword. 

Could we stick to type, and if people really want this in the future we can find a way to make it work with JSON Schema 2019-09 (and OAS 3.1)?

**Comment by nulltoken** [comment]
> This might actually go above and beyond. The feature request was for type, but this is type and format. Format is notoriously vague and I worry we're opening ourselves up to a maintainability nightmare, because JSON Schema 2019-09 has several new formats, and we're moving away from format being a validation keyword.

So my attempt at properly narrowing the scope in https://github.com/stoplightio/spectral/issues/570#issuecomment-570942127 was eventually ... completely useless 😜 

> Could we stick to type, and if people really want this in the future we can find a way to make it work with JSON Schema 2019-09 (and OAS 3.1)?

No problem. I'll dumb it down and should have something ready by tomorrow evening.

**Comment by rossmcdonald** [comment]
Sorry @nulltoken, I read 'format' as 'type' in your comment, so I'm most likely to blame for the extra effort here. 

After speaking with @philsturgeon, I also wasn't aware of how much 'format' was a can of worms when it comes to validation, so I agree limiting scope now will probably save us from a 'death by a thousand cuts' later as everyone has their own specific definition of each format.

**Comment by nulltoken** [comment]
@rossmcdonald No problem! I'm actually happy to reschedule all the life-lessons I could learn from 'formats' 😄 

@philsturgeon I've pushed a simpler version. Does it fit you better?

BTW, although I've tried to be cautious, it's possible some "format" related things may still hide in the corners. If you spot some, please add a comment so that I can clean that up.

File: src/functions/__tests__/typedEnum.test.ts [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    object
+
+    */
+    const testCases: Array<[string, unknown[], unknown]> = [
+      ['string', ['Hello', 'world!'], 12],
+      ['number', [-2147483648, 17.13], 'Hello'],
+      ['integer', [-2147483648, 17], 12.3],
+      ['boolean', [true, false], 1],
+      // array ?
+      // object ?
```
Comment (nulltoken) [inline src/functions/__tests__/typedEnum.test.ts]: @philsturgeon I now realize I've actually never met this use case.

It looks like (from my reading of both oas2 and oas3 specs) that an enum could also typed as  an `array` or an `object`.

Do you confirm this point?

File: docs/reference/functions.md [hunk 10]
```diff
@@ -264,3 +264,13 @@ example-value-or-externalValue:
       - externalValue
       - value
 ```
+
+## typedEnum
+
+Models described by both a `type` and an `enum` must only expose enum values that respect this type.
```
Comment (philsturgeon) [inline docs/reference/functions.md]:
```suggestion
When both a `type` and `enum` are defined for a property, the enum values must respect the type.
```

This makes me think this would be better off as a custom function for the oas ruleset. Afterall, this is oas/json schema specific, which means it doesn't really belong in core functions, right?

File: docs/reference/functions.md [hunk 11]
```diff
@@ -264,3 +264,13 @@ example-value-or-externalValue:
...
+
+<!-- title: functionOptions -->
+
+name                   | description                                                        | required?
+-----------------------|--------------------------------------------------------------------|----------
+reportingThreshold     | the maximum number of invalid values reported in the error message | yes
```
Comment (philsturgeon) [inline docs/reference/functions.md]: This seems unique to this rule, whats the idea here?

File: src/functions/__tests__/typedEnum.test.ts (hunk 9 above)
Comment (philsturgeon) [reply src/functions/__tests__/typedEnum.test.ts]: OAS 2 and 3 don't have any opinions of their own about what is or is not a valid instance for an enum, but, JSON Schema does: https://json-schema.org/draft/2019-09/json-schema-validation.html#enum

It says it can be used for anything, but I have never see anyone try to do`enum: [[0, 1], [2,3]]` and I don't think we want to waste our time figuring out how to make that happen right now. Let's stick to the cases you've mentioned here, and make sure we test no type (no type = any, so no error)

File: src/functions/__tests__/typedEnum.test.ts [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+      'should add an ellipsis at the end of the error message when the number of invalid enum values is greater than the reporting threshold "%s"',
+      (threshold, errorMessageTrail) => {
+        const schema = {
+          type: 'integer',
+          enum: ['a string!', 'and another one!', 'seriously?', 'you must be kidding me'],
+        };
+
+        expect(runTypedEnum(schema, threshold)).toEqual([
+          {
+            message: `Some enum values do not respect the specified type "integer": ${errorMessageTrail}`,
```
Comment (philsturgeon) [inline src/functions/__tests__/typedEnum.test.ts]: Could the error be reported for each invalid enum value instead of all of them? this might cut out the need for reportingThreshold.

File: test-harness/scenarios/typed-enum.oas3.scenario [hunk 13]
```diff
@@ -0,0 +1,33 @@
+====test====
+Identify enum values that do not respect the specified format
```
Comment (philsturgeon) [inline test-harness/scenarios/typed-enum.oas3.scenario]:
```suggestion
Identify enum values that do not respect the specified type
```

File: src/functions/__tests__/typedEnum.test.ts [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+    const schema = {
+      type: 'integer',
+      enum: [],
+    };
+
+    expect(runTypedEnum(schema, defaultReportingThreshold)).toBeUndefined();
+  });
+
+  describe('basic', () => {
+    test('should return undefined when all enum values respect the format', () => {
```
Comment (philsturgeon) [inline src/functions/__tests__/typedEnum.test.ts]:
```suggestion
    it('is undefined when all enum values respect the type', () => {
```

File: src/functions/__tests__/typedEnum.test.ts [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+      const schema = {
+        type: 'integer',
+        enum: [123, 456],
+      };
+
+      expect(runTypedEnum(schema, defaultReportingThreshold)).toBeUndefined();
+    });
+
+    test('should identify enum values that do not respect the format', () => {
```
Comment (philsturgeon) [inline src/functions/__tests__/typedEnum.test.ts]:
```suggestion
    it('identifies each enum value which does not respect the type', () => {
```

File: docs/reference/functions.md [hunk 16]
```diff
@@ -264,3 +264,8 @@ example-value-or-externalValue:
       - externalValue
       - value
 ```
+<!--
```
Comment (nulltoken) [inline docs/reference/functions.md]: @philsturgeon Switching to custom function means I can drop that, as we only document core functions, right?

File: package.json [hunk 17]
```diff
@@ -85,6 +85,7 @@
   },
   "devDependencies": {
     "@commitlint/config-conventional": "^8.3.4",
+    "@rollup/plugin-json": "^4.0.1",
```
Comment (nulltoken) [inline package.json]: Required since we're switching to a custom function model and ajv needs this to be bundled.

File: rollup.config.js [hunk 18]
```diff
@@ -23,19 +24,21 @@ for (const file of fs.readdirSync(targetDir)) {
 module.exports = functions.map(fn => ({
   input: fn,
   plugins: [
+    resolve({preferBuiltins: false}),
```
Comment (nulltoken) [inline rollup.config.js]: I'm a tad lost here.... and haven't found the proper way to tame "assert" and "os"

```
(!) Plugin node-resolve: preferring built-in module 'assert' over local alternative at 'H:\spectral\node_modules\assert\build\assert.js', pass 'preferBuiltins: fal
se' to disable this behavior or 'preferBuiltins: true' to disable this warning
(!) Plugin node-resolve: preferring built-in module 'os' over local alternative at 'os', pass 'preferBuiltins: false' to disable this behavior or 'preferBuiltins: 
true' to disable this warning
(!) Unresolved dependencies
https://rollupjs.org/guide/en/#warning-treating-module-as-external-dependency
assert (imported by dist\rulesets\oas\functions\oasTypedEnum.js,  assert?commonjs-external)
os (imported by node_modules\better-ajv-errors\node_modules\supports-color\index.js,  os?commonjs-external, node_modules\@babel\highlight\node_modules\supports-col
```

@P0lip Help? 🙏

File: rollup.config.js [hunk 19]
```diff
@@ -23,19 +24,21 @@ for (const file of fs.readdirSync(targetDir)) {
...
     commonjs({
       namedExports: {
         'node_modules/lodash/lodash.js': ['isObject', 'trimStart'],
         'node_modules/@stoplight/types/dist/index.js': ['DiagnosticSeverity'],
       },
     }),
+    json(),
     terser(),
   ],
+  // external: ["assert", "os"],
```
Comment (nulltoken) [inline rollup.config.js]: Should we make them external? Does it even make sense?

File: docs/reference/functions.md (hunk 16 above)
Comment (philsturgeon) [reply docs/reference/functions.md]: Correct, we document the OpenAPI rules and the function is just an implementation detail.

File: src/functions/__tests__/typedEnum.test.ts (hunk 9 above)
Comment (nulltoken) [reply src/functions/__tests__/typedEnum.test.ts]: >Let's stick to the cases you've mentioned here
@phil Done.

>make sure we test no type (no type = any, so no error)
Covered by an additional test case

File: src/functions/__tests__/typedEnum.test.ts (hunk 12 above)
Comment (nulltoken) [reply src/functions/__tests__/typedEnum.test.ts]: @phil Fixed. Multiple messages are now returned and reportingThreahold has been shown the door.

File: src/utils/prepareResults.ts [hunk 20]
```diff
@@ -18,6 +18,8 @@ export const defaultComputeResultFingerprint: ComputeFingerprintFunc = (rule, ha
 
   if (rule.source) id += rule.source;
 
+  if (rule.message) id += rule.message;
```
Comment (nulltoken) [inline src/utils/prepareResults.ts]: @marbemac I''ve had to patch the fingerprinting function which was too agressive fro this new test cases. Does that fit you?

File: src/functions/typedEnum.ts [hunk 21]
```diff
@@ -0,0 +1,40 @@
+import { isArray } from 'lodash';
+import { IFunction, IFunctionResult, IRule, RuleFunction } from '../types';
+import { schema } from './schema';
+
+export type TypedEnumRule = IRule<RuleFunction.TYPED_ENUM>;
+
+export const typedEnum: IFunction = (targetVal, opts, paths, otherValues): void | IFunctionResult[] => {
+  const { enum: enumValues, ...initialSchema } = targetVal;
+
+  if (!isArray(enumValues)) {
```
Comment (nulltoken) [inline src/functions/typedEnum.ts]: @philsturgeon We currenlty bail out if not an array. @P0lip mentioned that we may also report an malformed enum. However, shouldn't a malformed enum be identified by the `oas[2|3]-schema` rule?

File: src/utils/prepareResults.ts (hunk 20 above)
Comment (marbemac) [reply src/utils/prepareResults.ts]: Interesting - which test case(s) fail if this is not added? The fingerprint already takes into account the rule code (`typed-enum`) and the error range/path - is there a situation in which we would want two `typed-enum` errors reported at the same path? Or maybe I'm not understanding the underlying issue.

Comment (nulltoken) [reply src/utils/prepareResults.ts]: @marbemac The rule relies on a function which analyzes properties exposing both a `type` and an `enum`. The function return an array of messages. Each message points at an invalid enum value. 

```OpenAPI 3.x detected

{document}
 7:12  error  typed-enum  Enum value "Tap, tap, tap..." do not respect the specified type "integer".
 7:12  error  typed-enum  Enum value "Is thing thing on?" do not respect the specified type "integer".
 7:12  error  typed-enum  Enum value "You certainly wonder" do not respect the specified type "integer".
 7:12  error  typed-enum  Enum value "Why I gathered you all today..." do not respect the specified type "integer".

✖ 4 problems (4 errors, 0 warnings, 0 infos, 0 hints)
```

Without that change, only one error message would be returned above, as the returned results only differ by the message.

Comment (marbemac) [reply src/utils/prepareResults.ts]: Ah interesting - I guess the new question is do we want to do it this way? Versus a single error that indicates there are enum values that don't match.

```
7:12  error  typed-enum  Enum values "Tap, tap, tap...", "Is thing thing on?",  "You certainly wonder", and "Why I gathered you all today..." do not respect the specified types "integer".
```

Don't have a super strong opinion on this, but it would mean the assumption of one error per file location + error code is still a rule that holds true. Other potential benefit is that it results in less reporting noise since the occurrences would be grouped together instead of reported as distinct errors.

Will leave it to you and @P0lip / @philsturgeon to decide :).

Comment (nulltoken) [reply src/utils/prepareResults.ts]: @philsturgeon So, what's your vote on this? Revert to a reporting threshold (to prevent long list from generating a too long long error message)? Or stick with current implementation which may have the unexpected (good?) side effect of fixing #920 as well?

I understand @marbemac 's point in that concatenating would result in a terser reporting, although I wouldn't tag it as noise, as each error message would point at a different real issue.

Comment (philsturgeon) [reply src/utils/prepareResults.ts]: @nulltoken @marbemac I think it depends on how you view the rule conceptually. Are we reporting that the enum is invalid because it has some bad items listed in it? Or are we reporting that the array item is invalid? More often than not I see people using the muli line YAML synax:

```
1: type: boolean
2: enum: 
3: - true
4: - nonsense
```

I would expect to see error message for line 4 not line 2, because line 4 is the problem not line 2. 

if there are 3 problematic lines in the enum, I would expect 3 errors, because each problem is another thing to fix.

File: src/functions/typedEnum.ts (hunk 21 above)
Comment (philsturgeon) [reply src/functions/typedEnum.ts]: @nulltoken yeah I think this rule has a very specific target of "make sure the array items match the type defined" so other stuff should be another.

File: src/utils/prepareResults.ts (hunk 20 above)
Comment (marbemac) [reply src/utils/prepareResults.ts]: If we go w @philsturgeon's suggestion, then `range` would be different per reported error, and original fingerprint strategy should work as is.

However, it's been a while since I've been in this are of Spectral codebase, so I'm not sure how difficult it would be to report path/range at an array index resolution (`['myProp', 'enum', 1]`) rather than at a array property resolution (`['myProp', 'enum']`).

Comment (nulltoken) [reply src/utils/prepareResults.ts]: @philsturgeon @marbemac  Indeed, it makes much more sense to point at the value rather at the enum level. On it!

Comment (nulltoken) [reply src/utils/prepareResults.ts]: @philsturgeon @marbemac 
- The rule now reports the ranges of the offending values
- The change to the fingerprinting function has been reverted

File: src/rulesets/oas/__tests__/typed-enum.ts [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+            start: {
+              character: 10,
+              line: 10,
+            },
+            end: {
+              character: 28,
+              line: 10,
+            },
+          },
+          severity: 1,
```
Comment (P0lip) [inline src/rulesets/oas/__tests__/typed-enum.ts]:
```suggestion
          severity: DiagnosticSeverity.Warning,
```

File: src/rulesets/oas/__tests__/typed-enum.ts [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+            start: {
+              character: 10,
+              line: 8,
+            },
+            end: {
+              character: 21,
+              line: 8,
+            },
+          },
+          severity: 1,
```
Comment (P0lip) [inline src/rulesets/oas/__tests__/typed-enum.ts]:
```suggestion
          severity: DiagnosticSeverity.Warning,
```

File: src/rulesets/oas/__tests__/typed-enum.ts [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+      };
+
+      const results = await s.run(doc);
+
+      expect(results).toEqual([
+        {
+          code: 'typed-enum',
+          message: 'Enum value "a string!" does not respect the specified type "integer".',
+          path: ['definitions', 'Test', 'enum', '1'],
+          range: {
```
Comment (P0lip) [inline src/rulesets/oas/__tests__/typed-enum.ts]: Let's screw ranges here, as they are not really relevant.
```suggestion
          range: expect.any(Object),
```

File: src/rulesets/oas/__tests__/typed-enum.ts [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+            start: {
+              character: 8,
+              line: 7,
+            },
+            end: {
+              character: 19,
+              line: 7,
+            },
+          },
+          severity: 1,
```
Comment (P0lip) [inline src/rulesets/oas/__tests__/typed-enum.ts]:
```suggestion
          severity: DiagnosticSeverity.Warning,
```

File: src/rulesets/oas/__tests__/typed-enum.ts [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+            start: {
+              character: 8,
+              line: 9,
+            },
+            end: {
+              character: 26,
+              line: 9,
+            },
+          },
+          severity: 1,
```
Comment (P0lip) [inline src/rulesets/oas/__tests__/typed-enum.ts]: severity: DiagnosticSeverity.Warning,

File: src/rulesets/oas/__tests__/typed-enum.ts [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+      };
+
+      const results = await s.run(doc);
+
+      expect(results).toEqual([
+        {
+          code: 'typed-enum',
+          message: 'Enum value "a string!" does not respect the specified type "integer".',
+          path: ['components', 'schemas', 'Test', 'enum', '1'],
+          range: {
```
Comment (P0lip) [inline src/rulesets/oas/__tests__/typed-enum.ts]:
```suggestion
          range: expect.any(Object),
```

File: src/rulesets/oas/__tests__/typed-enum.ts [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+              line: 7,
+            },
+          },
+          severity: 1,
+        },
+        {
+          code: 'typed-enum',
+          message: 'Enum value "and another one!" does not respect the specified type "integer".',
+          path: ['definitions', 'Test', 'enum', '3'],
+          range: {
```
Comment (P0lip) [inline src/rulesets/oas/__tests__/typed-enum.ts]:
```suggestion
          range: expect.any(Object),
```

File: src/functions/typedEnum.ts [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+    return;
+  }
+
+  const { type } = initialSchema;
+
+  const rootPath = paths.target ?? paths.given;
+
+  return incorrectValues.map(bad => {
+    return {
+      message: `Enum value "${bad.val}" does not respect the specified type "${type}".`,
```
Comment (P0lip) [inline src/functions/typedEnum.ts]:
```suggestion
      message: `Enum value \`${bad.val}\` does not respect the specified type \`${type}\`.`,
```
I believe we use grave accents in most cases now.

File: src/functions/typedEnum.ts [hunk 30]
```diff
@@ -0,0 +1,41 @@
+import { isArray } from 'lodash';
+import { IFunction, IFunctionResult, IRule, RuleFunction } from '../types';
+import { schema } from './schema';
+
+export type TypedEnumRule = IRule<RuleFunction.TYPED_ENUM>;
+
+export const typedEnum: IFunction = (targetVal, opts, paths, otherValues): void | IFunctionResult[] => {
+  const { enum: enumValues, ...initialSchema } = targetVal;
+
+  if (!isArray(enumValues)) {
```
Comment (P0lip) [inline src/functions/typedEnum.ts]:
```suggestion
  if (!Array.isArray(enumValues)) {
```

File: src/functions/typedEnum.ts [hunk 31]
```diff
@@ -0,0 +1,41 @@
+import { isArray } from 'lodash';
```
Comment (P0lip) [inline src/functions/typedEnum.ts]:
```suggestion
```

File: src/functions/formattedEnum.ts (hunk 7 above)
Comment (nulltoken) [reply src/functions/formattedEnum.ts]: Fixed. You're right.

FWIW, I don't even remember why I switched over to lodash... :-/

File: src/functions/formattedEnum.ts (hunk 8 above)
Comment (nulltoken) [reply src/functions/formattedEnum.ts]: Fixed

File: src/rulesets/oas/__tests__/typed-enum.ts (hunk 24 above)
Comment (nulltoken) [reply src/rulesets/oas/__tests__/typed-enum.ts]: Fixed

File: src/rulesets/oas/__tests__/typed-enum.ts (hunk 25 above)
Comment (nulltoken) [reply src/rulesets/oas/__tests__/typed-enum.ts]: Fixed

File: src/rulesets/oas/__tests__/typed-enum.ts (hunk 26 above)
Comment (nulltoken) [reply src/rulesets/oas/__tests__/typed-enum.ts]: Fixed

File: src/rulesets/oas/__tests__/typed-enum.ts (hunk 27 above)
Comment (nulltoken) [reply src/rulesets/oas/__tests__/typed-enum.ts]: Fixed

File: src/rulesets/oas/__tests__/typed-enum.ts (hunk 28 above)
Comment (nulltoken) [reply src/rulesets/oas/__tests__/typed-enum.ts]: Fixed

**Comment by nulltoken** [comment]
@P0lip All required changes have been fixed.

**Comment by nulltoken** [comment]
@philsturgeon It looks like the Check Markdown Links failed because it cannot resolve https://offset.earth/stoplightinc

Sweet irony!

image

And.. indeed, it's down

image

----

image

File: src/utils/prepareResults.ts (hunk 20 above)
Comment (marbemac) [reply src/utils/prepareResults.ts]: Very nice! Moving forward, IF we're ever in another situation where we're considering expanding the fingerprint function, it might indicate a separate issue :). I'm hopeful we can hold true to the "rule code + range/path" combo is unique in a result set.

**Comment by marbemac** [comment]
Ha - looks like the offset earth site is back up - just kicked off a re-run of those checks. 🤞
